extension/js/common/message-renderer.ts

```diff
@@ -0,0 +1,782 @@
+              name: attachment.name,
+              type: attachment.type,
+              length: attachment.getData().length,
+              data: attachment.getData(),
+              cid: attachment.cid, // todo: do we need it?
+            })
+          );
+        }
+      }
+    }
+    if (!sigResult?.match && signedContentInDecryptedData) {
+      const plaintext = signedContentInDecryptedData.rawSignedContent;
+      const sigText = signedContentInDecryptedData.signature.getData().toUtfStr();
+      const verify = (verificationPubs: string[]) => MsgUtil.verifyDetached({ plaintext, sigText, verificationPubs });
+      const newSigResult = await verify(verificationPubs);
+      return await this.decideDecryptedContentFormattingAndRender(
+        signerEmail,
+        verificationPubs,
+        plaintext,
+        isEncrypted,
+        newSigResult,
+        renderModule,
+        this.getRetryVerification(signerEmail, verify)
+      );
+    }
+    renderModule.separateQuotedContentAndRenderText(decryptedContent, isHtml);
+    await MessageRenderer.renderPgpSignatureCheckResult(renderModule, sigResult, Boolean(signerEmail), retryVerification);
+    if (renderableAttachments.length) {
+      renderModule.renderInnerAttachments(renderableAttachments, isEncrypted);
+    }
+    renderModule.resizePgpBlockFrame();
+    renderModule.setTestState('ready');
+    return isEncrypted ? { publicKeys } : {};
+  };
+
+  private relayAndStartProcessing = async (
+    relayManager: RelayManager,
+    loaderContext: LoaderContextBindInterface,
+    factory: XssSafeFactory,
+    frameId: string,
+    printMailInfo: PrintMailInfo | undefined,
+    senderEmail: string | undefined,
+    cb: (renderModule: RenderInterface, frameId: string) => Promise<{ publicKeys?: string[]; needPassphrase?: string[] }>
+  ) => {
+    const renderModule = relayManager.createRelay(frameId);
+    loaderContext.bind(frameId, relayManager);
+    if (printMailInfo) {
+      renderModule.setPrintMailInfo(printMailInfo);
+    }
+    cb(renderModule, frameId)
+      .then(async result => {
+        const appendAfter = $(`iframe#${frameId}`); // todo: late binding? won't work
+        // todo: how publicKeys and needPassphrase interact?
+        for (const armoredPubkey of result.publicKeys ?? []) {
+          appendAfter.after(factory.embeddedPubkey(armoredPubkey, this.isOutgoing(senderEmail)));
+        }
+        while (result.needPassphrase) {
+          // todo: queue into a dictionary?
+          await PassphraseStore.waitUntilPassphraseChanged(this.acctEmail, result.needPassphrase);
+          renderModule.clearErrorStatus();
+          renderModule.renderText('Decrypting...');
+          result = await cb(renderModule, frameId);
+        }
+        // todo: wait for renderModule completion: the queue has been actually flushed,
+        // and then remove the frame from relayManager.frames
+        // Something like:
+        // await relayManager.waitForCompletion(renderModule)
```

Comment:
I think it'll make sense to clean up resources connected with a msgId as soon as all pgp_block frames are rendered (setTestState('ready') message is queued and queue is flushed).
We don't want the content-script internal variables to bloat, do we?

Reply:
Bloat would indeed be problematic. It happened in the past, some people leave the tab open for months.

If this is difficult to implement, then a simple alternative would be to delete resources that are > 24 hours old.

Reply:
Implemented 24 hours expiration